stan/math/opencl/matrix_cl.hpp

```diff
@@ -303,6 +303,40 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
     }
   }
 
+  /**
+   * Constructor for the matrix_cl that
+   * creates a copy of the Eigen Map of Eigen Matrix of doubles
+   * on the OpenCL device.
+   * Regardless of `partial_view`, whole matrix in the map
+   * is stored.
+   *
+   * @tparam T type of data in the \c Eigen \c Matrix
+   * @param A the \c Eigen \c Map of the Eigen Matrix
+   * @param partial_view which part of the matrix is used
+   *
+   * @throw <code>std::system_error</code> if the
+   * matrices do not have matching dimensions
+   */
+  template <int R, int C>
+  explicit matrix_cl(Eigen::Map<Eigen::Matrix<T, R, C>>& A,
```

Comment:
I think we can just do an Eigen::MatrixBase<Derived> constructor here instead of a map constructor

Reply:
I tried that, but had issues then accessing `.data()` without doing stuff like deep copies so I resorted to a separate constructor. Will try again.

Reply:
I got it working with Eigen::Ref, is that fine with you?

So the signature of the constructor is now

```cpp
explicit matrix_cl(const Eigen::Ref<const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>>& A,
                     matrix_cl_view partial_view = matrix_cl_view::Entire)
      : rows_(A.rows()), cols_(A.cols()), view_(partial_view) {
```

Reply:
Nice! Yep that's cool with me